npm/private/test/snapshots/npm_defs.bzl

```diff
@@ -2931,6 +2951,385 @@ def npm_link_all_packages(name = "node_modules", imported_links = []):
+
+    if accessing_package == "examples/nextjs":
+        packages_to_validate.append("tailwindcss")
+
+    if accessing_package == "js/private/worker/src":
+        packages_to_validate.append("tslib")
+
+    if accessing_package == "npm/private/test":
+        packages_to_validate.append("typescript")
+
+    if accessing_package == "":
+        packages_to_validate.append("typescript")
+
+    if accessing_package == "js/private/worker/src":
+        packages_to_validate.append("typescript")
+
+    if accessing_package == "npm/private/test":
+        packages_to_validate.append("unused")
+
+    if accessing_package == "examples/npm_package/packages/pkg_a":
+        packages_to_validate.append("uuid")
+
+    if accessing_package == "examples/npm_package/packages/pkg_b":
+        packages_to_validate.append("uuid")
+
+    if accessing_package == "examples/npm_package/packages/pkg_d":
+        packages_to_validate.append("uuid")
+
+    if accessing_package == "examples/npm_deps":
+        packages_to_validate.append("uvu")
+
+    if accessing_package == "npm/private/test":
+        packages_to_validate.append("webpack-bundle-analyzer")
+
+    if accessing_package == "examples/webpack_cli":
+        packages_to_validate.append("webpack-cli")
+
+    if accessing_package == "examples/webpack_cli":
+        packages_to_validate.append("webpack")
+
+
+    # Validate each package
+    for package_name in packages_to_validate:
+        if not _check_package_visibility(accessing_package, package_name):
+            fail("""
+Package visibility violation:
+
+  Package: {}
+  Requested by: {}
+
+This package is not visible from your location.
+Check the package_visibility configuration in your npm_translate_lock rule.
+
+For more information, see: https://docs.aspect.build/rules/aspect_rules_js/docs/npm_translate_lock#package_visibility
+""".format(package_name, accessing_package))
+
+def _check_package_visibility(accessing_package, package_name):
+    """Check if accessing_package can access package_name"""
+
+    # Get visibility rules for this package
+    visibility_rules = _get_package_visibility_rules(package_name)
+
+    # Check each visibility rule
+    for rule in visibility_rules:
+        if rule == "//visibility:public":
+            return True
+
+        # Package-specific access: //packages/foo:__pkg__
+        if rule == "//" + accessing_package + ":__pkg__":
+            return True
+
+        # Subpackage access: //packages/foo:__subpackages__
+        if rule.endswith(":__subpackages__"):
+            rule_package = rule[2:-16]  # Remove "//" and ":__subpackages__"
+            if accessing_package.startswith(rule_package + "/") or accessing_package == rule_package:
+                return True
+
+        # Target-specific access: //packages/foo:target
+        if rule.startswith("//" + accessing_package + ":"):
+            return True
+
+    return False
+
+def _get_package_visibility_rules(package_name):
```

Comment:
Most of this method is consistent and not generated, can it be imported from somewhere instead of generated? Then it doesn't need the ugly template string in `npm/private/npm_translate_lock_generate.bzl` etc.

I think you'd just have to make `_NPM_PACKAGE_VISIBILITY` a param?